Add prettier exclusions to our configuration

AnimiVulpis opened the pull request:
This will turn off all rules that will be obsolete as soon as we start
using prettier to format our front end code.

This should not be merged before the initial **prettyfication** is done.

If you think we should create a separate configuration please feel free and explain why you think this would be a good idea.

Commit:
Add prettier exclusions to our configuration

This will turn off all rules that will be obsolete as soon as we start
using prettier to format our front end code.

AnimiVulpis commented:
For details about our roadmap see https://github.com/signavio/frontend-technical-debt/pull/1

Commit:
Add .editorconfig file

Commit:
Rewrite eslint-config-signavio without changing a lot

* Delete `index.js` (replaced with `.eslintrc.js`)
* Delete `.eslintrc.json` (replaced with `.eslintrc.js`)
* Bump the version (`package.json`)
* Change the `main` attribute in the `package.json`
* Add npm scripts for
 * Linting
 * Formatting
* Add myself as contributor
* Update and replace dependencies
* Update `yarn.lock` file

Commit:
Update README.md

Please pay attention to the **Open Questions that need an answer**
section

AnimiVulpis commented:
I put some effort into updating the `eslint-config-signavio` ESLint configuration.

Please take some time and look over the changes.

Please pay special attention to the **Open Questions that need an answer** section in the `README.md`

jfschwarz commented on .editorconfig:
```diff
@@ -0,0 +1,16 @@
+
+[package.json]
+indent_style = space
+indent_size = 2
```
why not also 4?

jbreckel replied:
this file will be parsed and rewritten by `npm`/`yarn` with 2 spaces. so we would constantly have the whole file as diff.

AnimiVulpis replied:
Correct.
This can be changed as soon as npm5 is used by all developers
(npm5 makes changes to the `package.json` indentation preserving)

jfschwarz commented on .eslintrc.js:
```diff
@@ -0,0 +1,76 @@
+            },
+        ],
+        'no-mixed-operators': [
+            'off',
```
why did you switch this off? I read your explanation in the README but I don't get it.

AnimiVulpis replied:
Because if I kept it on prettier would disagree with eslint, but I wasn't sure how important this rule is for others

jfschwarz commented on .eslintrc.js:
```diff
@@ -0,0 +1,76 @@
+        'no-mixed-operators': [
+            'off',
+            {
+                allowSamePrecedence: true,
```
if we really have to switch this off, it would not make sense to define these options anylonger...

AnimiVulpis replied:
As soon as there are no objections to removing these we can. For the sake of the pull request it seemed easier

jfschwarz commented on README.md:
```diff
@@ -31,3 +30,102 @@ After that create a `.eslintrc.json` file and add the following:
+
+### Open Questions that need an answer
+
+- `no-mixed-operators` should we turn this rule on somehow?
```
would be nice... but maybe also not thaaat urgent.

AnimiVulpis replied:
I remembered some discussion around this rule.
From my point of view I agree

jfschwarz replied:
ok, so if it not easy to configure, then I'm fine with disabling it

jbreckel reviewed (changes requested):
I would rather make this a major version update. Apart from that LGTM

jfschwarz commented on README.md:
```diff
@@ -31,3 +30,102 @@ After that create a `.eslintrc.json` file and add the following:
+### Open Questions that need an answer
+
+- `no-mixed-operators` should we turn this rule on somehow?
+- [`no-use-before-define`](http://eslint.org/docs/rules/no-use-before-define) was turned to `error` (old config does not work anymore, any change needed?)
```
why was turning to error necessary?

AnimiVulpis replied:
Because I couldn't figure out what the old config did and thought: Closer to the original is better

AnimiVulpis replied:
Because I couldn't figure out what the old config did and thought: Closer to the original is better

jfschwarz replied:
we had some good reason for disabling it. of course feel free to challenge this, but it should be a separate discussion. it's not related to prettier, right?

jfschwarz commented on README.md:
```diff
@@ -31,3 +30,102 @@ After that create a `.eslintrc.json` file and add the following:
+
+- `no-mixed-operators` should we turn this rule on somehow?
+- [`no-use-before-define`](http://eslint.org/docs/rules/no-use-before-define) was turned to `error` (old config does not work anymore, any change needed?)
+- were `"jsx-a11y/onclick-has-focus": "off"` and `"jsx-a11y/onclick-has-role": "off"` replaced with `"jsx-a11y/interactive-supports-focus": "error"`?
```
I'd say, upgrading to latest airbnb config is definitely good. There's no need to discuss any of the changes coming in with this update here. If we find some of the new rules to cause troubles for us, we can still discuss changing or disabling them in separate PR discussions.

AnimiVulpis replied:
Good approach, like it

jfschwarz commented on README.md:
```diff
@@ -31,3 +30,102 @@ After that create a `.eslintrc.json` file and add the following:
 However please always make a good point, why you want to introduce a change.
 Simply "not liking" something doesn't count.
+
+# Changelog
```
I'd prefer to have the changelog in a separate file. Blows up the readme otherwise

AnimiVulpis replied:
Agreed. Will change that

cristovaoth commented on .eslintrc.js:
```diff
@@ -0,0 +1,76 @@
+module.exports = {
+    extends: ['airbnb', 'prettier', 'prettier/flowtype', 'prettier/react'],
```
Why does this need to be so complicated?? Why do keep having airbnb there? Again, airbnb makes a lot of style/formatting assumptions.
Why not:

1- Come from eslint recommended
2- turn off what ever doesnt make sense
3- extend prettier and prettier/react?

AnimiVulpis replied:
Because airbnb has also a lot of rules regarding flow and react that people seem to like and I would prefer not to waste a lot of time trying to come up with configs for those (and a few more e.g. a11y)

cristovaoth replied:
Isn't there any other solution? If prettier is getting real popular - which it lookes like - isn't there a popular eslint-config to match it?

This sounds like duck-taping things together, and also, looking at the config file starts to get complicated.

Eslint is just a list of rules, and a list of plugins. Should be easier

cristovaoth replied:
If we all assume this is a temporary solution: temporarily like in actively searching for a better set rules, that do not include any styling/format rules ok.

Otherwise, I oppose to keeping airbnb there.

jfschwarz replied:
> If prettier is getting real popular - which it lookes like - isn't there a popular eslint-config to match it?

Let's keep our eyes open. Maybe there will be new popular eslint config adjusted to work together with prettier, maybe airbnb will adjust theirs.

The airbnb ruleset focuses on stuff that help developers avoid common pitfalls. so it's complementary to prettier. they also have stuff for formatting obviously, but it certainly would make sense for the airbnb guys to simply remove all formatting rules overlapping with prettier from their ruleset.

For now, we have to do this tiny bit of duck-taping, but I also do not see a real pain this might cause. It's still just a (concatenated) list of rules–last rule wins.

cristovaoth reviewed (changes requested):
I want to do a proper review of all these changes next week.

AnimiVulpis commented:
Action items until now:
- [x] Make this change a major version bump
- [x] Move the Changelog into a separate file
- [x] Delete turned off rules if no objections arise

Commit:
Make this PR a major version update

Commit:
Split changelog from readme

Commit:
Change _special_ `prettier` rules

And add section about those 3 rules to `CHANGELOG.md`

AnimiVulpis commented:
Updated according to comments

jbreckel commented on README.md:
```diff
@@ -31,3 +30,7 @@ After that create a `.eslintrc.json` file and add the following:
+
+# Changelog
+
+See `CHANGELOG.md`
```
maybe add a link to the file? ( I am not sure how this works )

AnimiVulpis replied:
Will search and then do

AnimiVulpis replied:
Done

jbreckel commented on CHANGELOG.md:
```diff
@@ -0,0 +1,81 @@
+# v2.1.0
```
Please update to major version ( as package.json )

AnimiVulpis replied:
Ups, forgot that one

AnimiVulpis replied:
Done

jbreckel reviewed (changes requested):
nice work, found just 2 minor things

Commit:
Fix version in and link to CHANGELOG.md

(from README.md)

AnimiVulpis commented:
Two reviews to go 🤞

AnimiVulpis commented:
Last one

Commit:
Restore some rule changes by the original signavio config

With the comments explaining them.

Except for the flow warnings; No information here

AnimiVulpis commented:
Added some old rules back based on feedback from @jfschwarz

jfschwarz commented:
v3.0.0 is now published
